nss-day-cohort-38/bangazon-ecommerce-api-iris-station
# Kp image upload

Fixes #14 #26

# Fetching Instructions (Back End)
```shell
git fetch --all
git checkout kp-image-upload
git pull
```
Pillow had to be installed for picture handling reasons...
```
pip install -r requirements.txt
```
Models have been changed to allow images to be uploaded (as opposed to just being url paths), so you'll have to migrate those:
```
python manage.py makemigrations
python manage.py migrate
```

Product Fixtures have been updated to utilize new images in the media folder, so install them:
```
python manage.py loaddata */fixtures/product.json
```

# Fetching Instructions (Front End)
```shell
git fetch --all
git checkout kp-image-upload
git pull
```

# Description
Please include a summary of the change and which issue is fixed.

- Users can upload images when creating new products
- Those images are stored in the /media/` folder on the back end api repo (note: they are gitignored though)
- Products retrieved through list or detail will display those images

**Note:** images url paths on products will no longer work. Any previously created products (besides those updated with the fixtures) will need to either 
1. have their `image_path` field deleted (you can do this in tableplus) 
2. be deleted altogether (sorry!) or 
3. have the `image_path` field updated manually with the same filename as a corresponding image file that you'll have to place in the `/media/` folder
# Testing Process

1. In the browser, click "Sell a Product" on the navbar
1. Create a product normally, but upload an image file (!!!) 

After Submission, Confirm:
- [ ] the product appears on the list view with an image
- [ ] the product detail page also shows the image
- [ ] that image file is in the `/media/` folder of your back end repo

1. In the browser, click "Sell a Product" on the navbar
1. Create a product normally, but do not upload an image file

Confirm:
- [ ] you are not allowed to complete the form unless you've uploaded an image

1. In the browser, click "Sell a Product" on the navbar
1. Create a product normally, but upload a file larger than 5MB (like [this one](https://stocksnap.io/photo/marsh-field-LQCNP30RBB))

Confirm:
- [ ] you are prompted with an error and not allowed to submit

# List of changes (back end only)
- Urls have had a static route added 
- Settings have had media route and url added 
- Product Fixtures have been updated to utilize local media files 
- Local image files added for aforementioned fixtures (see media folder)
- Product Models updated to use ImageField
- Product create view updated to handle request.FILES 
- updated product model image_path field to allow blank or null values 
- added an if statement to the create products view, to allow a user to post a product without an image and have it result as null

# Checklist:
